File: Contests/LeetCode W406/4.java
import java.util.*;

class Solution {
    public long minimumCost(int m, int n, int[] horizontalCut, int[] verticalCut) {
        Arrays.sort(horizontalCut);
        reverseArray(horizontalCut);
        Arrays.sort(verticalCut);
        reverseArray(verticalCut);

        long a = 0, b = 0; 
        long hp = 1, vp = 1;
        long rst = 0;

        while (a < horizontalCut.length && b < verticalCut.length) {
            if (horizontalCut[(int)a] > verticalCut[(int)b]) {
                rst += (long)horizontalCut[(int)a] * vp;
                hp++;
                a++;
            } else {
                rst += (long)verticalCut[(int)b] * hp;
                vp++;
                b++;
            }
        }

        while (a < horizontalCut.length) {
            rst += (long)horizontalCut[(int)a] * vp;
            a++;
        }

        while (b < verticalCut.length) {
            rst += (long)verticalCut[(int)b] * hp;
            b++;
        }

        return rst;
    }
    private void reverseArray(int[] arr) {
        int left = 0, right = arr.length - 1;
        while (left < right) {
            int temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
    }
}

File: Contests/LeetCode W406/4.cpp
#include <vector>
#include <iostream>
#include <vector>
#include <__algorithm/pstl_sort.h>
class Solution {
public:
    long long minimumCost(int m, int n, vector<int>& horizontalCut, vector<int>& verticalCut) {
        sort(horizontalCut.rbegin(), horizontalCut.rend());
        sort(verticalCut.rbegin(), verticalCut.rend());
        
        long long h = 0, v = 0; // Use long long to avoid overflow
        long long horizontalPieces = 1, verticalPieces = 1;
        long long totalCost = 0;
        
        while (h < horizontalCut.size() && v < verticalCut.size()) {
            if (horizontalCut[h] > verticalCut[v]) {
                totalCost += (long long)horizontalCut[h] * verticalPieces;
                horizontalPieces++;
                h++;
            } else {
                totalCost += (long long)verticalCut[v] * horizontalPieces;
                verticalPieces++;
                v++;
            }
        }
        
        while (h < horizontalCut.size()) {
            totalCost += (long long)horizontalCut[h] * verticalPieces;
            h++;
        }
        
        while (v < verticalCut.size()) {
            totalCost += (long long)verticalCut[v] * horizontalPieces;
            v++;
        }
        
        return totalCost;

    }
};
